File: nodejs/controllers/url.js
/**
 * Database connection
 */
const db = require('../db').get('gonodejsperf');

/**
 * Create indexes
 */
db.collection('url').createIndex({ url: 1 }, { unique: true });

/**
 * Generate pseudorandom string
 * @param {Number} len Length of the expected string
 */
const randomstring = (len) => {
    let alphabet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_';
    let str = '';
    for (let i = 0; i < len; i++) {
        str += alphabet.charAt(Math.floor(Math.random() * alphabet.length));
    }
    return str;
}

/**
 * Get list of all shorened urls
 */
module.exports.getList = (req, res) => {
    db.collection('url').find({}).project({
        _id: false,
        url: true,
        target: true,
        used: true,
        lastUse: true,
    }).toArray((err, data) => {
        if (err) {
            return res.json({
                status: 'ERROR',
                msg: err.toString(),
                data: null,
            });
        }

        // send ok response
        res.json({
            status: 'OK',
            msg: 'Full URL list successfully retrieved.',
            data: data,
        });
    });
};

/**
 * Do the redirect
 */
module.exports.redir = (req, res) => {
    db.collection('url').findOneAndUpdate({
        url: req.params.url,
    }, {
        $inc: {
            used: 1,
        },
        $set: {
            lastUse: new Date(),
        },
    }, {
        projection: {
            _id: false,
            target: true,
        },
        returnOriginal: false,
    }, (err, data) => {
        if (err || !data.value) {
            return res.redirect('/');
        }

        // redirect to the original url
        res.redirect(data.value.target);
    });
};

/**
 * Create new shortened URL
 * @param req.body.target The target URL
 */
module.exports.create = (req, res) => {
    // create document for save
    let doc = {
        url: randomstring(6),
        target: req.body.target,
        used: 0,
        lastUse: null,
        created: new Date(),
        modified: new Date(),
        password: randomstring(32),
    };

    // insert to database
    db.collection('url').insertOne(doc, (err, data) => {
        if (err) {
            return res.json({
                status: 'ERROR',
                msg: err.toString(),
                data: null,
            });
        }

        // send ok response
        res.json({
            status: 'OK',
            msg: 'URL was shortened.',
            data: data.ops[0],
        });
    });
};

/**
 * Update existing URL's target
 * @param req.body.password Password given on url creation
 * @param req.body.target New target URL
 */
module.exports.update = (req, res) => {
    db.collection('url').findOneAndUpdate({
        url: req.params.url,
        password: req.body.password,
    }, {
        $set: {
            modified: new Date(),
            target: req.body.target,
        },
    }, {
        returnOriginal: false,
    }, (err, data) => {
        if (err) {
            return res.json({
                status: 'ERROR',
                msg: err.toString(),
                data: null,
            });
        }

        // test data (if no = password wrong)
        if (!data.lastErrorObject.updatedExisting) {
            return res.json({
                status: 'ERROR',
                msg: 'The password you entered is not valid.',
                data: null,
            });
        }

        // send ok response
        res.json({
            status: 'OK',
            msg: 'The target of the URL was updated successfully.',
            data: data.value,
        });
    });
};

/**
 * Delete existing shorened URL
 * @param req.params.url The shortened url id to delete
 */
module.exports.delete = (req, res) => {
    db.collection('url').deleteOne({
        url: req.params.url,
    }, (err) => {
        if (err) {
            return res.json({
                status: 'ERROR',
                msg: err.toString(),
                data: null,
            });
        }

        // send ok response
        res.json({
            status: 'OK',
            msg: 'Shortened URL was deleted.',
        });
    });
};

File: go/controllers/url.go
package controllers

import (
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
	"github.com/jansvabik/go-nodejs-performance/go/data"
)

// GetList gets all URLs from database and sends it to client
func GetList(w http.ResponseWriter, r *http.Request) {
	URLs, err := data.GetList()
	if err != nil {
		writeErrorResponse(w, err.Error(), http.StatusInternalServerError)
		return
	}

	writeOkResponse(w, "Full URL list successfully retrieved.", URLs)
}

// Redir does the redirect to the shortened URL target
func Redir(w http.ResponseWriter, r *http.Request) {
	URLID := mux.Vars(r)["url"]
	URL, err := data.Use(URLID)
	if err != nil {
		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
		return
	}

	http.Redirect(w, r, URL.Target, http.StatusTemporaryRedirect)
}

// Create creates new document and saves it in database
func Create(w http.ResponseWriter, r *http.Request) {
	// decode the URL body
	var d data.URL
	err := json.NewDecoder(r.Body).Decode(&d)
	if err != nil {
		http.Error(w, "Cannot decode your JSON. You should submit JSON like {\"target\": \"https://some.url/\"}.", http.StatusBadRequest)
		return
	}

	// try to create the URL
	doc, err := data.Create(d.Target)
	if err != nil {
		writeErrorResponse(w, err.Error(), http.StatusInternalServerError)
		return
	}

	writeOkResponse(w, "URL was shortened.", doc)
}

// Update updates the target of specified document in database
func Update(w http.ResponseWriter, r *http.Request) {
	URLID := mux.Vars(r)["url"]

	// decode the URL body
	var d data.URL
	err := json.NewDecoder(r.Body).Decode(&d)
	if err != nil {
		http.Error(w, "Cannot decode your JSON. You should submit JSON like {\"target\": \"https://some.url/\", \"password\": \"YourSecretPasswordFromCreation\"}.", http.StatusBadRequest)
		return
	}

	// try to delete the URL
	doc, err := data.Update(URLID, d.Target, d.Password)
	if err != nil {
		if err.Error() == "mongo: no documents in result" {
			writeErrorResponse(w, "The password you entered is not valid.", http.StatusInternalServerError)
		} else {
			writeErrorResponse(w, err.Error(), http.StatusInternalServerError)
		}
		return
	}

	writeOkResponse(w, "The target of the URL was updated successfully.", doc)
}

// Delete deletes the specified document from database
func Delete(w http.ResponseWriter, r *http.Request) {
	URLID := mux.Vars(r)["url"]

	// try to delete the URL
	err := data.Delete(URLID)
	if err != nil {
		writeErrorResponse(w, err.Error(), http.StatusInternalServerError)
		return
	}

	writeOkResponse(w, "Shortened URL was deleted.", nil)
}
